### PR TITLE
Core dump on ignition off during registering 20 apps

### PR DESCRIPTION
**Pull request by RomanReznichenkoLuxoft**
Fixes #[3857](https://github.com/smartdevicelink/sdl_core/issues/3857)

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Manual test using steps in issue description

### Summary
Problems I fixed.

The first problem. Аn object of this class AppExtension was not given externally without protection. Therefore, I will change the return type from Extensions () to DataAccessor <std :: list <AppExtensionPtr>>. Where the desired mutex is captured.

Second problem. ApplicationManagerImpl had unprotected deletion of objects used in other threads. Also, I found a duplicate of the is_stopping method. I deleted this one and left the IsStopping method.

Third problem. ResumeCtrlImpl captured the mutex queue_lock_, which belongs to it, and the mutex from ApplicationManagerImpl, when calling the application(..) method. This created a deadlock, and not a secure use of the application object. So I added getting all the apps and working with this list.

Fourth problem. It consists of two different crash of the program. The first crash when remove mobile_request_list_. The second is when deleting mobile requests from waiting_for_response_. This was due to the use of smart pointers in different threads. They are copied into threads, so we have to delete them in the thread in which we finish working with it. Since these requests are processed in a separate thread, I delete them there if the core stops working during the processing of these requests. Also added a check to see if the core stopped working while adding such requests to the list.

The fifth problem. This is a drop while processing data in one thread and deleting it in another thread. I created a separate mutex to protect this data. Because using the existing one, we do not invest in the time allotted to the timer to perform certain work.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by RomanReznichenkoLuxoft** [comment]
@theresalech This PR is ready for Livio review.

File: src/components/application_manager/test/request_controller/request_controller_test.cc [hunk 1]
```diff
@@ -258,11 +241,23 @@ TEST_F(RequestControllerTestClass, IsLowVoltage_SetOnWakeUp_FALSE) {
 }
 
 TEST_F(RequestControllerTestClass, AddMobileRequest_SetValidData_SUCCESS) {
```
Comment (ShobhitAd) [inline src/components/application_manager/test/request_controller/request_controller_test.cc:243]: This unit test seems to fail sometimes
```
      Start 17: request_controller_test
17/36 Test #17: request_controller_test .............Child aborted***Exception:   1.16 sec
97% tests passed, 1 tests failed out of 36

Total Test time (real) =  39.75 sec

The following tests FAILED:
	 17 - request_controller_test (Child aborted)
```

Seeing this error message in the Testing/Temporary/LastTest.log

```
[ RUN      ] RequestControllerTestClass.AddMobileRequest_SetValidData_SUCCESS
/home/shobhit/sdl/CORE/test/sdl_core/src/components/application_manager/test/request_controller/request_controller_test.cc:254: Failure
Actual function call count doesn't match EXPECT_CALL(app_mngr_, IsStopping())...
         Expected: to be called once
           Actual: never called - unsatisfied and active
pure virtual method called
terminate called without an active exception
<end of output>
Test time =   1.16 sec
----------------------------------------------------------
Test Failed.
```

**Review by ShobhitAd — CHANGES_REQUESTED** [review]

File: src/components/application_manager/test/request_controller/request_controller_test.cc (hunk 1 above)
Comment (ShobhitAd) [reply src/components/application_manager/test/request_controller/request_controller_test.cc:243]: Hey @RomanReznichenkoLuxoft, the test seems to fail every 1/50 runs. No special conditions, I saw the test fail the first time I ran `make test` and just re-ran `make test` repeatedly until I encountered the issue again.

Comment (RomanReznichenkoLuxoft) [reply src/components/application_manager/test/request_controller/request_controller_test.cc:243]: Hi @ShobhitAd. Thanks, I also detected locally. Now I analyze the tests and life of objects. I will let you know when there is a fix.

**Pull request closed without merging**